### PR TITLE
Ensure color codes don't span newlines

### PR DESCRIPTION
**Pull request by robx**
This is meant to address #346. It's not at all polished, but should be technically "done". Do let me know if you agree with this change and the approach in principle -- I'd be more than happy to make any desired changes to make the code nicer.

### CODE REVIEW

File: hspec-core/src/Test/Hspec/Core/Formatters/Internal.hs [hunk 1]
```diff
@@ -225,9 +226,27 @@ clearTransientOutput = do
 
 -- | Append some output to the report.
 write :: String -> FormatM ()
-write s = do
+write str = do
+  useColor <- getConfig formatConfigUseColor
```
Comment (robx) [inline hspec-core/src/Test/Hspec/Core/Formatters/Internal.hs:230]: an alternative would be to check `useColor` in `withColor_`, not sure which is nicer

**Comment by sol** [comment]
@robx this looks great!  I'll rebase + resolve conflicts and merge.

Sorry for the delay.

File: hspec-core/src/Test/Hspec/Core/Formatters/Internal.hs (hunk 1 above)
Comment (sol) [reply hspec-core/src/Test/Hspec/Core/Formatters/Internal.hs:230]: I think that is what I would do.  I may refactor it.

File: hspec-core/src/Test/Hspec/Core/Formatters/Internal.hs [hunk 2]
```diff
@@ -259,21 +278,12 @@ htmlSpan :: String -> FormatM a -> FormatM a
...
-      (when useColor $ hSetSGR h [Reset])
-
-      -- run action
-      (runStateT action st)
+withColor_ color action = do
+  oldColor <- gets stateColor
+  modify (\state -> state { stateColor = Just color })
+  x <- action
+  modify (\state -> state { stateColor = oldColor })
+  return x
```
Comment (sol) [inline hspec-core/src/Test/Hspec/Core/Formatters/Internal.hs:286]: I really like this.  It's great how we get arbitrary color nesting for free with this approach :clap:

**Comment by sol** [comment]
@robx I squashed your changes into a single commit (to facilitate `git bisect` if needed at some point in the future).  I preserved your authorship, but I changed the commit message to match the title of this PR. I also applied some refactoring on top of it:

1. Move the `useColor` check to `withColor_` https://github.com/hspec/hspec/pull/643/commits/8c4b40796c34b5f94289984ad0db34df958b4cda
2. Only colorize non-whitespace chunks https://github.com/hspec/hspec/pull/643/commits/d76e3efc0874b961cc4b37f1236010ffc7469210
3. With (2) in place, it felt natural to first chunk the output string and the do a `mapM_` over it (instead of `doLines`) https://github.com/hspec/hspec/pull/643/commits/3a4cff2e2c418b394d91267040a0d2a47fde97c7
4. Having unit tests for (3) I thought it's ok to get rid of the test in `RunnerSpec` and instead only test that `write` works for one simple case (we already know that chunking works correctly after all) https://github.com/hspec/hspec/pull/643/commits/9c1b47be3c26ad396c024239407906f3ec8f5cf3

The result is on https://github.com/hspec/hspec/pull/643.

I think this is pretty awesome now.  I really like the approach you took.  Somehow I lacked inspiration on how to tackle this.  Thanks a bunch!  Really much appreciated!

**Pull request closed without merging**